subprojects/language-native/src/main/java/org/gradle/language/cpp/plugins/CppBasePlugin.java

```diff
@@ -85,25 +91,32 @@ public void execute(ComponentAwareRepository componentAwareRepository) {
             public void execute(final CppBinary binary) {
                 final Names names = Names.of(binary.getName());
 
-                CppCompile compile = tasks.create(names.getCompileTaskName("cpp"), CppCompile.class);
-                compile.includes(binary.getCompileIncludePath());
-                compile.source(binary.getCppSource());
-                if (binary.isDebuggable()) {
-                    compile.setDebuggable(true);
-                } else {
-                    compile.setOptimized(true);
-                }
-                compile.getObjectFileDir().set(buildDirectory.dir("obj/" + names.getDirName()));
-
-                DefaultNativePlatform currentPlatform = new DefaultNativePlatform("current");
-                compile.setTargetPlatform(currentPlatform);
-
+                String language = "cpp";
+                final DefaultNativePlatform currentPlatform = new DefaultNativePlatform("current");
                 // TODO - make this lazy
-                NativeToolChain toolChain = modelRegistry.realize("toolChains", NativeToolChainRegistryInternal.class).getForPlatform(currentPlatform);
-                compile.setToolChain(toolChain);
+                final NativeToolChain toolChain = modelRegistry.realize("toolChains", NativeToolChainRegistryInternal.class).getForPlatform(currentPlatform);
+
+                Callable<List<File>> systemIncludes = new Callable<List<File>>() {
+                    @Override
+                    public List<File> call() throws Exception {
+                        PlatformToolProvider platformToolProvider = ((NativeToolChainInternal) toolChain).select(currentPlatform);
+                        if (platformToolProvider instanceof SystemIncludesAwarePlatformToolProvider) {
+                            return ((SystemIncludesAwarePlatformToolProvider) platformToolProvider).getSystemIncludes();
+                        }
+                        return ImmutableList.of();
+                    }
+                };
+
+                CppCompile compile = tasks.create(names.getCompileTaskName(language), CppCompile.class);
+                configureCompile(compile, binary, currentPlatform, toolChain, systemIncludes);
+                compile.getObjectFileDir().set(buildDirectory.dir("obj/" + names.getDirName()));
 
                 ((DefaultCppBinary)binary).getObjectsDir().set(compile.getObjectFileDir());
 
+                Depend depend = tasks.create(names.getTaskName("depend") + StringUtils.capitalize(language), Depend.class);
```

Comment:
Maybe just add this to `Names` as `getDependTaskName(language)`?

Reply:
I added `getDependTaskName(language)` to `Names`.